src/typeEvaluator/typeEvaluate.ts

```diff
@@ -1047,316 +1149,60 @@ function isPrimitiveTypeNode(node: TypeNode): node is PrimitiveTypeNode {
-              // eslint-disable-next-line max-depth
-              if (right.of.value === undefined) {
-                return undefined
-              }
-              patterns = patterns.concat(matchAnalyzePattern(right.of.value))
-            }
-            if (right.of.type === 'union') {
-              // eslint-disable-next-line max-depth
-              for (const node of right.of.of) {
-                // eslint-disable-next-line max-depth
-                if (node.type === 'string') {
-                  // eslint-disable-next-line max-depth
-                  if (node.value === undefined) {
-                    return undefined
-                  }
-                  patterns = patterns.concat(matchAnalyzePattern(node.value))
-                }
-
-                // eslint-disable-next-line max-depth
-                if (node.type !== 'string') {
-                  return false
-                }
-              }
-            }
-          }
-          return matchText(tokens, patterns)
-        }
-        case '<': {
-          if (isPrimitiveTypeNode(left) && isPrimitiveTypeNode(right)) {
-            if (left.value === undefined || right.value === undefined) {
-              return undefined
-            }
-            return left.value < right.value
-          }
-
-          return undefined
-        }
-        case '<=': {
-          if (isPrimitiveTypeNode(left) && isPrimitiveTypeNode(right)) {
-            if (left.value === undefined || right.value === undefined) {
-              return undefined
-            }
-            return left.value <= right.value
-          }
-
-          return undefined
-        }
-        case '>': {
-          if (isPrimitiveTypeNode(left) && isPrimitiveTypeNode(right)) {
-            if (left.value === undefined || right.value === undefined) {
-              return undefined
-            }
-            return left.value > right.value
-          }
-
-          return undefined
-        }
-        case '>=': {
-          if (isPrimitiveTypeNode(left) && isPrimitiveTypeNode(right)) {
-            if (left.value === undefined || right.value === undefined) {
-              return undefined
-            }
-            return left.value >= right.value
-          }
-
-          return undefined
-        }
-
-        default: {
-          return undefined
-        }
-      }
-    }
-
-    case 'Not': {
-      const result = resolveCondition(expr.base, scope)
-      // check if the result is undefined or false. Undefined means that the condition can't be resolved, and we should keep the node
-      return result === undefined ? undefined : result === false
+    case '>=': {
+      return left.value >= right.value
     }
-
-    case 'Group': {
-      return resolveCondition(expr.base, scope)
-    }
-
     default: {
-      return undefined
+      throw new Error(`unknown comparison operator ${opcall}`)
     }
   }
 }
 
+function nodeMatch(node: TypeNode): boolean | undefined {
```

Comment:
Maybe rename this to `booleanValue(node: TypeNode)` and then clarify that it's intended to be used in the cases where we coerce a value to a boolean (i.e. inside filter + object splat).

Note that in a filter then every other type is actually considered `false`: https://spec.groq.dev/draft/#sel-FARJLCFHFFAACJAuyC. So actually, all of `object`, `number`, `array`, `null` can be considered `false` in this context.